Refine arm pl011 upstream

SgrrZhf opened the pull request:
In this PR, I add all available uart_pl011 devices on fvp_baser_aemv8r platforms.

To make use of all available uart_pl011 devices, I refine creating the device instance code of uart_pl011. 
- The old way creates uart_pl011 devices one by one only supports 2 uart_pl011 devices or fewer.
- The new way creates uart_pl011 devices via macro 'DT_INST_FOREACH_STATUS_OKAY' and without the number limit of  uart_pl011.

At last, I remove 'CONFIG_UART_PL011_PORT0=y' from defconfig files because we can select the uart_pl011 device by setting 'status=ok' in the device tree file.

I can't remove 'UART_PL011_PORT0' from Kconfig.pl011 file because this option is used by v2m_musca_s1 board and I don't have this board. Hope someone can remove them fully.
https://github.com/zephyrproject-rtos/zephyr/blob/1d189487b4d5c6cba0be8c2856e738ec44e6562d/boards/arm/v2m_musca_s1/pinmux.c#L44

SgrrZhf committed:
board: arm64: fvp-baser-aemv8r: add all available uart nodes

fvp-baser-aemv8r has four pl011_uart devices and all of then have
been added in this patch.

only uart0 and uart1 are enabled as default in fvp_baser_aemv8r.dts

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
driver: uart_pl011: refine creating device instance code

Old code only create one pl011 device instance though there are two or
more pl011 device defined in devicetree. This patch can fix this issue.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
boards: remove options CONFIG_UART_PL011_PORT0

Remove option 'CONFIG_UART_PL011_PORT0' from four boards's defconfig
file.

- fvp_baser_aemv8r_aarch32
- fvp_base_revc_2xaemv8a
- fvp_baser_aemv8r
- qemu_cortex_a53

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

povergoing reviewed (dismissed):
LGTM, so the `CONFIG_UART_PL011_PORTn` will be defined automatically according to the uart nums and status in dts?

SgrrZhf commented:
> LGTM, so the `CONFIG_UART_PL011_PORTn` will be defined automatically according to the uart nums and status in dts?

No, Only two boards below select 'CONFIG_UART_PL011_PORTn' in the Kconfig.defconfig file. 
- v2m_musca_b1
https://github.com/zephyrproject-rtos/zephyr/blob/d92e3b88bacc6e6fc1b6222421ef4accdf2d574e/boards/arm/v2m_musca_b1/Kconfig.defconfig#L25
 
- v2m_musca_s1
https://github.com/zephyrproject-rtos/zephyr/blob/d92e3b88bacc6e6fc1b6222421ef4accdf2d574e/boards/arm/v2m_musca_s1/Kconfig.defconfig#L25

Four boards below don't need 'CONFIG_UART_PL011_PORTn' anymore.

- fvp_baser_aemv8r_aarch32
- fvp_base_revc_2xaemv8a
- fvp_baser_aemv8r
- qemu_cortex_a53

carlocaione commented on dts/arm64/fvp-aemv8r/fvp-aemv8r.dtsi:
```diff
@@ -79,5 +79,35 @@
 			label = "UART_0";
 			clocks = <&uartclk>;
 		};
+
+		uart1: uart@9c0a0000 {
```
Is this valid also for `fvp_base_revc_2xaemv8a`? Can you extend that as well?

SgrrZhf replied:
Yes, of course. I can do that.

carlocaione commented on drivers/serial/uart_pl011.c:
```diff
@@ -459,132 +459,65 @@ void pl011_isr(const struct device *dev)
-		    &pl011_cfg_port_1, PRE_KERNEL_1,
-		    CONFIG_SERIAL_INIT_PRIORITY,
-		    &pl011_driver_api);
+#define PL011_IRQ_CONFIG_FUNC_BODY(n, prop, i) \
+	{ \
+		IRQ_CONNECT(DT_IRQ_BY_IDX(n, i, irq), \
+			DT_IRQ_BY_IDX(n, i, priority), \
+			pl011_isr, \
+			DEVICE_DT_GET(n), \
+			0); \
+		irq_enable(DT_IRQ_BY_IDX(n, i, irq)); \
+	}
```
nit: please use a common format to align the `\` at the end of the line.

SgrrZhf replied:
Done

SgrrZhf committed:
boards: arm64: enable uart interrupt driven feature on two boards

- fvp_baser_aemv8r_smp
- fvp_base_revc_2xaemv8a

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
boards: arm64: fvp_base_revc_2xaemv8a: add all available uart nodes

fvp_base_revc_2xaemv8a has four pl011_uart devices and all of then have
been added in this patch.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

povergoing reviewed (approved):
refresh +1
